### ISSUE TITLE
Number of epoch for training the teacher model (SwinT+MulTi and SwinT+Multi+SelfD)?

### ISSUE THREAD

**Issue by maxboels** [IMSY-DKFZ/self-distilled-swin#12]
Hi,

Should we train the teacher with 3 or 20 epochs? In the README, it says:

Train SwinT+MulTi and SwinT+Multi+SelfD
`bash runs/run.sh target_size=131 **save_epoch=3** exp=SwinT+MultiT
`
but in the paper:
"Self-Distillation The concept of self-distillation was achieved by training a
teacher Swin transformer on one-hot encoded hard labels for **20 epochs**, with
a batch size of 64, ..."

Thank you

**Comment by Amiiney** [comment]
Hi, in the paper, we trained the teacher for 20 epochs and we saved the weights based on the best validation mAP which was around the 3rd epoch (no performance improvement after this epoch). We recommend training the teacher for 3 epochs only to get similar results to the ones reported in the paper.

**Comment by maxboels** [comment]
Ok, thanks for the clarification and fast response

**Comment by Amiiney** [comment]
Welcome! Just to clarify, only the teacher should be trained for 3 epochs, the student needs to be trained for 40 epochs. This should be handled by the bash script runs/run.sh as mentioned in the readme. EDIT: Student for 40 epochs

**Comment by maxboels** [comment]
Yes, I can confirm the student is training over 40 epochs.

**Comment by maxboels** [comment]
One more question, is there a script to save the frame embeddings for all folds using the different weights?

**Comment by Amiiney** [comment]
We don't have a script that runs automatically to generate frame embedding, we might add it in the future. However, you can use generate.py that generates soft-labels to generate frame embeddings instead by removing the classifier during the model loading:

```
    def load_feature_extractor(self):
        """
        Loads the model as a feature extractor by removing the classification head.

        This is specifically for cases where CFG.sequential is True, in order to extract image embeddings
        """
        weights_path = os.path.join(
            self.CFG.output_dir,
            f"checkpoints/fold{self.fold}_{self.CFG.model_name[:8]}_{self.CFG.target_size}_{self.CFG.exp}.pth",
        )

         model = TripletModel(self.CFG, pretrained=False).to(self.CFG.device)
  

        pretrained_dict = torch.load(weights_path)["model"]
        classifier_keys = [
            "model.head.fc.weight",
            "model.head.fc.bias",
            "model.head.weight",
            "model.head.bias",
            "model.classifier.weight",
            "model.classifier.bias",
        ]
        pretrained_dict = {
            k: v for k, v in pretrained_dict.items() if k not in classifier_keys
        }
        model.load_state_dict(pretrained_dict)
        self.model = model
        print("Feature extractor loaded successfully!")
```

This will automatically save the frame embeddings in the csv file, you might need to adapt[ this part of the code](https://github.com/IMSY-DKFZ/self-distilled-swin/blob/bb4e2884c81b72ffd343a8b6babee6d546db51e7/generate.py#L66C29-L66C30) to save them in your preferred format.

**Comment by maxboels** [comment]
Great, thanks for suggesting a script.
I was able to get `CholecT50 mAP: 0.3408` on the official cross-validation splits with 3 epochs for the teacher, and 40 epochs for the student using the "SwinT+MultiT" model. I removed the `null_verb` and `null_target` classes which might affect the multi-task training.

**Comment by Amiiney** [comment]
So, SwinT+MultiT achieves 0.3408 on CholecT50. How about SwinT+MultiT+SelfD?

Did you remove the null_verb and null_target classes from training or only from the metric calculation? In our paper, we calculated the metric using all 100 classes during cross-validation; however, test set results were computed without the null classes."

**Comment by maxboels** [comment]
Sorry, I meant 0.3408  with the "SwinT+MultiT+SelfD" model.

I removed those classes from the multi-task labels but kept it in the triplets eval.

**Comment by Amiiney** [comment]
ah ok, so I understand that you are using the extact same setup in the configuration, the only difference is removing the null classes? I think this should not have a big impact on the performance. How did you compute the results?

**Comment by maxboels** [comment]
Yes, I don't think it's a bad thing to remove those classes since we are looking for multiple objects presence in frames with a sigmoid binary cross entropy. I didn't change the updated evaluation script from your last push in the new branch.
